77812

SAM19991 opened the pull request.

SAM19991 committed:
77812

SAM19991 committed:
77812

SAM19991 committed:
77812

SAM19991 committed:
77812

Hitman-nn commented on src/main/java/faang/school/projectservice/controller/VacancyController.java:
```diff
@@ -0,0 +1,73 @@
+        );
+    }
+
+    @PostMapping("/{vacancyId}/close")
```
Почему Post для операции изменения?

SAM19991 replied:
Тут я посчитал , что Post будет уместен , т.к. это именно операция закрыть, а не чистое обновление поля. Или я ошибаюсь?

Hitman-nn replied:
GET — получение информации об объекте (ресурсе).
POST — создание нового объекта (ресурса).
PUT — полная замена объекта (ресурса) на обновленную версию.
PATCH — частичное изменение объекта (ресурса).
DELETE — удаление информации об объекте (ресурсе).

и даже в самом ТЗ:
"**Обновление вакансий:**
Вакансия может быть обновлена только пользователем с ролью OWNER или MANAGER
Закрыть вакансию можно только в случае, если найдено требуемое количество кандидатов. У выбранных кандидатов должна быть назначена соответствующая роль на проекте."

И сам же просто меняешь поле.

Hitman-nn commented on src/main/java/faang/school/projectservice/service/VacancyServiceImpl.java:
```diff
@@ -0,0 +1,111 @@
+
+    @Override
+    public VacancyDto updateVacancy(long projectId, long vacancyId, VacancyDto vacancyDto) {
+        checkOwnerOrManager(projectId);
+        Vacancy existingVacancy = loadVacancy(vacancyId);
+        validateVacancyOwnership(existingVacancy, projectId);
+        Vacancy vacancyToUpdate = vacancyMapper.toEntity(vacancyDto);
+        vacancyToUpdate.setProject(existingVacancy.getProject());
+        vacancyToUpdate.setId(existingVacancy.getId());
+        vacancyToUpdate.setStatus(existingVacancy.getStatus());
+        Vacancy savedVacancy = vacancyRepository.save(vacancyToUpdate);
+        return vacancyMapper.toDto(savedVacancy);
```
А зачем так сложно? В чем задумка?
existingVacancy.setName(vacancyDto.что-то());

SAM19991 replied:
исправил

Hitman-nn commented on src/main/java/faang/school/projectservice/mapper/VacancyMapper.java:
```diff
@@ -0,0 +1,23 @@
+@Mapper(componentModel = "spring", uses = CandidateMapper.class)
+public interface VacancyMapper {
+
+    VacancyMapper INSTANCE = Mappers.getMapper(VacancyMapper.class);
```
Это зачем тут? какой смысл?

SAM19991 replied:
Забыл удалить, проверял кое-что.

Hitman-nn replied:
опять как-то странно делаешь.
+ попробуй через mapper сделать и @target в ней

Hitman-nn commented on src/main/java/faang/school/projectservice/service/VacancyServiceImpl.java:
```diff
@@ -0,0 +1,111 @@
+
+    @Override
+    public List<VacancyDto> getVacanciesByProjectId(long projectId, String positionFilter, String nameFilter) {
+        return vacancyRepository.findAll().stream()
+                .filter(vacancy -> vacancy.getProject().getId().equals(projectId))
+                .filter(vacancy -> positionFilter == null || vacancy.getPosition().name().equalsIgnoreCase(positionFilter))
+                .filter(vacancy -> nameFilter == null || vacancy.getName().contains(nameFilter))
+                .map(vacancyMapper::toDto)
+                .toList();
```
Если расширять фильтр, что будет? Может стоит попробовать применить паттерн стратегия.

SAM19991 replied:
Что-то сделал , но не уверен , что это верный паттерн стратегия. Создал  подпакет service/filter, в  нём завёл интерфейс VacancyFilter с методом boolean test(Vacancy vacancy) и для каждого типа фильтра сделал класс-реализацию в том же пакете. А тут сделал сборку и применение этих фильтров.

Hitman-nn commented on src/main/java/faang/school/projectservice/service/VacancyServiceImpl.java:
```diff
@@ -0,0 +1,111 @@
+        }
+    }
+
+    @Override
+    public VacancyDto createVacancy(long projectId, VacancyDto vacancyDto) {
+        checkOwnerOrManager(projectId);
+        var project = projectRepository.findById(projectId)
+                .orElseThrow(() -> new DataValidationException("Project not found"));
+        Vacancy vacancy = vacancyMapper.toEntity(vacancyDto);
+        vacancy.setProject(project);
+        vacancy.setStatus(VacancyStatus.OPEN);
+        Vacancy savedVacancy = vacancyRepository.save(vacancy);
+        return vacancyMapper.toDto(savedVacancy);
+    }
```
Не увидел не тут, не далее:
"Проверить, что добавляемые кандидаты не являются участниками проекта."

Поясни пожалуйста как и где это делается

SAM19991 replied:
Сейчас такой проверки нет, её нужно добавить в методы)). В условиях задания этой проверки не было, а самому додуматься у меня еще не хватает опыта.

SAM19991 replied:
Добавил

Hitman-nn replied:
ну это вроде в задании было) я это не придумал)

miked4king replied:
вижу что используется проверка в методе обновления, метод validateCandidatesNotProjectMembers
а при создании стоит добавить данную проверку?

Hitman-nn reviewed (commented):
В общем все понравилось)

SAM19991 committed:
Changed Post to Put in the closeVacancy method.

Simplifying updateVacancy so as not to recreate the entire entity via mapper, but to change only the necessary fields in the existingVacancy.

Implemented (not sure) the “Strategy” pattern for filtering.

Fixed minor things and removed unnecessary things.

Hitman-nn commented on src/main/java/faang/school/projectservice/service/filter/VacancyFilter.java:
```diff
@@ -0,0 +1,7 @@
+import faang.school.projectservice.model.Vacancy;
+
+public interface VacancyFilter {
+    boolean test(Vacancy vacancy);
```
почему название такое странное?
смотрел это https://faang-school.com/courses/4jnzmndg/9nzw6rrd

SAM19991 replied:
Теперь как в ролике

Hitman-nn commented on src/main/java/faang/school/projectservice/service/VacancyServiceImpl.java:
```diff
@@ -0,0 +1,144 @@
+            validateCandidatesNotProjectMembers(newCandidates, projectId);
+        }
+
+        existingVacancy.setName(vacancyDto.getName());
+        existingVacancy.setDescription(vacancyDto.getDescription());
+        existingVacancy.setPosition(vacancyDto.getPosition());
+        existingVacancy.setCount(vacancyDto.getCount());
```
В принципе пойдет. 
Можно покрасивее еще mapper сделать и target в ней

SAM19991 replied:
Done

Hitman-nn commented on src/main/java/faang/school/projectservice/service/VacancyServiceImpl.java:
```diff
@@ -0,0 +1,144 @@
+                })
+                .map(vacancyMapper::toDto)
+                .toList();
+    }
```
а если по определенным фильтрам захочу фильтровать или ту задуманно, чтоб только все или ничего?
Посомотри воршоп по фильтрам https://faang-school.com/courses/4jnzmndg/9nzw6rrd
Просто штука реально удобная, прикольная, да и паттерн всегда пригодится

SAM19991 replied:
Вроде получилось как в ролике

SAM19991 committed:
1. Улучшил маппер:
    - добавил метод updateVacancy с @MappingTarget для безопасного обновления сущности
    - проставил ignore = true для полей, которые не должны обновляться (id, audit поля, связи)
    - упростил обновление vacancy в сервисе

2. Переделал фильтры:
    - переименовал в более понятные названия с постфиксом VacancyFilter
    - сделал их Spring компонентами через @Component
    - переработал интерфейс VacancyFilter (добавил методы isApplicable и apply)

3. Обновил VacancyServiceImpl:
    - добавил инжект списка фильтров
    - сделал более элегантную работу с фильтрами через Stream API
    - упростил обновление сущностей через новый метод маппера

4. Поправил тесты под новую логику

miked4king commented on src/main/java/faang/school/projectservice/controller/VacancyController.java:
```diff
@@ -0,0 +1,73 @@
+    @GetMapping("/ping")
+    public ResponseEntity<String> ping() {
+        return ResponseEntity.ok("pong");
+    }
```
???
видимо забыл убрать?

miked4king commented on src/main/java/faang/school/projectservice/controller/VacancyController.java:
```diff
@@ -0,0 +1,73 @@
+import java.util.List;
+
+@RestController
+@RequestMapping("/projects/{projectId}/vacancies")
```
давай будем придерживаться лучших практик и добавим версионирование /api/v1/projects/{projectId}/vacancies

miked4king commented on src/main/java/faang/school/projectservice/controller/VacancyController.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    @PostMapping
+    public ResponseEntity<VacancyDto> createVacancy(
```
а для чего везде добавил ResponseEntity?
вроде везде статус 200 возвращаешь, он по умолчанию будет.
В целом нет смысла добавлять ResponseEntity если у тебя статус 200 по умолчанию и ты не добавляешь кастомные заголовки, давай уберем

miked4king commented on src/main/java/faang/school/projectservice/controller/VacancyController.java:
```diff
@@ -0,0 +1,73 @@
+    @PutMapping("/{vacancyId}")
+    public ResponseEntity<VacancyDto> updateVacancy(
+            @PathVariable long projectId,
+            @PathVariable long vacancyId,
+            @RequestBody VacancyDto vacancyDto) {
```
в vacancyDto есть поле id, оно идентично полю vacancyId ?
если да, тогда зачем мы его передаем в пути?

miked4king commented on src/main/java/faang/school/projectservice/exeption/DataValidationException.java:
```diff
@@ -0,0 +1,8 @@
+package faang.school.projectservice.exeption;
```
кажется есть ошибка в названии пакета)

SAM19991 replied:
Done

miked4king commented on src/main/java/faang/school/projectservice/mapper/VacancyMapper.java:
```diff
@@ -0,0 +1,31 @@
+    @Mapping(source = "dto.status", target = "status")
+    Vacancy toEntity(VacancyDto dto);
+
+    @Mapping(target = "id", ignore = true)
+    @Mapping(target = "project", ignore = true)
+    @Mapping(target = "candidates", ignore = true)
+    @Mapping(target = "createdAt", ignore = true)
+    @Mapping(target = "updatedAt", ignore = true)
+    @Mapping(target = "createdBy", ignore = true)
+    @Mapping(target = "updatedBy", ignore = true)
```
так много полей в игноре стоят, а мы не можем просто сменить необходимое для нас поле (или сколько полей нам нужно) через сеттер?

SAM19991 replied:
Можем и изначально так и было, Данила посоветовал сделать так для красоты в методе, чтобы не было много строк. Ну или я его не так понял)))

miked4king commented on src/main/java/faang/school/projectservice/service/VacancyServiceImpl.java:
```diff
@@ -0,0 +1,140 @@
+    private final TeamMemberRepository memberRepo;
+    private final List<VacancyFilter> filterList;
+
+    public void checkOwnerOrManager(long projectId) {
```
почему метод public? в интерфейсе не видел его, наверное стоит на private изменить, если используется только в рамках данного метода.

Лучше чтобы в начале класса были методы, которые ты переопределяешь у интерфейса.
после них уже private методы, давай поправим

miked4king commented on src/main/java/faang/school/projectservice/service/VacancyServiceImpl.java:
```diff
@@ -0,0 +1,140 @@
+    private final VacancyMapper vacancyMapper;
+    private final ProjectRepository projectRepository;
+    private final UserContext userContext;
+    private final TeamMemberRepository memberRepo;
```
а тут зажали пару букв))
давай тоже назовем memberRepository

SAM19991 replied:
Done

miked4king commented on src/main/java/faang/school/projectservice/service/VacancyServiceImpl.java:
```diff
@@ -0,0 +1,140 @@
+
+        if (!hasRightRole) {
+            throw new DataValidationException(
+                    "Not allowed to perform this operation, only for owner and project manager"
```
будет здорово обогатить сообщение идентификатором, по которому мы искали

miked4king commented on src/main/java/faang/school/projectservice/service/VacancyServiceImpl.java:
```diff
@@ -0,0 +1,140 @@
+    @Override
+    public VacancyDto createVacancy(long projectId, VacancyDto vacancyDto) {
+        checkOwnerOrManager(projectId);
+        var project = projectRepository.findById(projectId)
```
давай в рамках буткемпа будем типизировать переменные, убери пж var (и если есть в других местах тоже)

miked4king commented on src/main/java/faang/school/projectservice/service/VacancyServiceImpl.java:
```diff
@@ -0,0 +1,140 @@
+    public VacancyDto createVacancy(long projectId, VacancyDto vacancyDto) {
+        checkOwnerOrManager(projectId);
+        var project = projectRepository.findById(projectId)
+                .orElseThrow(() -> new DataValidationException("Project not found"));
```
тоже было бы здорово id добавить в сообщение об ошибке.
добавь в других сообщениях тоже пж

miked4king commented on src/main/java/faang/school/projectservice/service/VacancyServiceImpl.java:
```diff
@@ -0,0 +1,140 @@
+
+        List<CandidateDto> newCandidates = vacancyDto.getCandidates();
+        if (newCandidates != null && !newCandidates.isEmpty()) {
+            validateCandidatesNotProjectMembers(newCandidates, projectId);
```
думаю лучше подойдет checkCandidatesNotProjectMembers

SAM19991 replied:
Done

miked4king commented on src/main/java/faang/school/projectservice/service/filter/NameVacancyFilter.java:
```diff
@@ -0,0 +1,22 @@
+    public Stream<Vacancy> apply(Stream<Vacancy> vacancyStream, String position, String name) {
+        return vacancyStream
+                .filter(vacancy ->
+                        vacancy.getName().toLowerCase().contains(name.toLowerCase()));
```
точно contains нужен?
кажется можем воспользоваться equalsIgnoreCase

SAM19991 replied:
В контексте поиска вакансий, думаю, лучше был бы  `contains`, так как это дает более гибкий поиск. Пользователь может искать по части названия вакансии.

miked4king commented on src/main/java/faang/school/projectservice/service/filter/NameVacancyFilter.java:
```diff
@@ -0,0 +1,22 @@
+    public Stream<Vacancy> apply(Stream<Vacancy> vacancyStream, String position, String name) {
+        return vacancyStream
+                .filter(vacancy ->
+                        vacancy.getName().toLowerCase().contains(name.toLowerCase()));
```
давай еще форматирование поправим на
return vacancyStream.filter(vacancy ->
         vacancy.getName().toLowerCase().contains(name.toLowerCase()));

miked4king commented on src/main/java/faang/school/projectservice/service/filter/PositionVacancyFilter.java:
```diff
@@ -0,0 +1,22 @@
+    public Stream<Vacancy> apply(Stream<Vacancy> vacancyStream, String position, String name) {
+        return vacancyStream
+                .filter(vacancy ->
+                        vacancy.getPosition().name().equalsIgnoreCase(position));
```
вот тут используешь как раз equalsIgnoreCase)
и форматирование поправим по аналогии с прошлым комментом

miked4king commented on src/main/java/faang/school/projectservice/service/filter/NameVacancyFilter.java:
```diff
@@ -0,0 +1,22 @@
+package faang.school.projectservice.service.filter;
```
я бы еще вынес пакет filter на уровень выше.
он как будто бы не относится к сервису

SAM19991 replied:
Done

SAM19991 committed:
- Сделал метод checkOwnerOrManager приватным для лучшей инкапсуляции
- Добавил проверку кандидатов при создании вакансии
- Заменил var на явное указание типов
- Улучшил сообщения об ошибках:
  - Добавил идентификаторы пользователя и проекта в checkOwnerOrManager
  - Добавил информацию о требуемом и текущем количестве в closeVacancy
- Реорганизовал структуру кода:
  - Переместил приватные методы в конец класса
  - Добавил проверку validateVacancyOwnership в closeVacancy
- Упростил VacancyServiceImpl.updateVacancy используя прямые сеттеры вместо маппера
- Улучшил форматирование PositionVacancyFilter для лучшей читаемости

SAM19991 committed:
- Обновлён base path для версионирования API (/api/v1/projects/{projectId}/vacancies).
- Убраны лишние ResponseEntity, для использования стандартного 200 статуса.
- Для обновления используется vacancyId в пути, а поле id возвращается в ответе.

miked4king commented on src/main/java/faang/school/projectservice/service/VacancyServiceImpl.java:
```diff
@@ -0,0 +1,155 @@
+        Stream<Vacancy> vacancyStream = vacancyRepository.findAll().stream()
+                .filter(vacancy -> vacancy.getProject().getId().equals(projectId));
+
+        if (filterList != null) {
```
а зачем нам данная проверка?
у тебя если реализаций интерфейса не будет то приложение не должно подняться, т.к не найдет бинов, давай уберем

miked4king commented on src/main/java/faang/school/projectservice/service/VacancyServiceImpl.java:
```diff
@@ -0,0 +1,155 @@
+        }
+
+        return vacancyStream
+                .map(vacancyMapper::toDto)
```
.map(vacancyMapper::toDto) - можно не переносить на новую строку

SAM19991 replied:
Так Влад всегда пишет в видео уроках.

SAM19991 committed:
файле VacancyServiceImpl , в методе getVacanciesByProjectId убрал лишнюю проверку filterList на null
